phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -1087,116 +1087,140 @@ private static PTable deserializeTable(byte[] b) {
+                                        mutations.add(del);
+                                        // Since we're replaying existing mutations, it makes no sense to write them to the wal
+                                        del.setDurability(Durability.SKIP_WAL);
+                                    }
+                                    del.addDeleteMarker(cell);
                                 }
-                                del.addDeleteMarker(cell);
                             }
+                            if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+                                checkForRegionClosing();
+                                commitBatchWithRetries(region, mutations, blockingMemstoreSize);
+                                uuidValue = ServerCacheClient.generateId();
+                                mutations.clear();
+                            }
+                            rowCount++;
                         }
-                        if (ServerUtil.readyToCommit(mutations.size(), mutations.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
-                            checkForRegionClosing();
-                            commitBatchWithRetries(region, mutations, blockingMemstoreSize);
-                            uuidValue = ServerCacheClient.generateId();
-                            mutations.clear();
-                        }
-                        rowCount++;
+
+                    } while (hasMore && rowCount < pageSizeInRows);
+                    if (!mutations.isEmpty()) {
+                        checkForRegionClosing();
+                        commitBatchWithRetries(region, mutations, blockingMemstoreSize);
                     }
-                    
-                } while (hasMore);
-                if (!mutations.isEmpty()) {
-                    checkForRegionClosing();
-                    commitBatchWithRetries(region, mutations, blockingMemstoreSize);
+                }
+            } catch (IOException e) {
+                hasMore = false;
+                LOGGER.error("IOException during rebuilding: " + Throwables.getStackTraceAsString(e));
+                throw e;
+            } finally {
+                if (!hasMore) {
+                    region.closeRegionOperation();
                 }
             }
-        } catch (IOException e) {
-            LOGGER.error("IOException during rebuilding: " + Throwables.getStackTraceAsString(e));
-            throw e;
-        } finally {
-            region.closeRegionOperation();
+            byte[] rowCountBytes = PLong.INSTANCE.toBytes(Long.valueOf(rowCount));
+            final Cell aggKeyValue = PhoenixKeyValueUtil.newKeyValue(UNGROUPED_AGG_ROW_KEY, SINGLE_COLUMN_FAMILY,
+                    SINGLE_COLUMN, AGG_TIMESTAMP, rowCountBytes, 0, rowCountBytes.length);
+            results.add(aggKeyValue);
+            rowCount = 0;
```

Comment:
It seems that rowCount needs not to be a member variable. Shall we change it to a local variable?

Reply:
I will change it

Reply:
Done